### PR TITLE
Add elasticsearch_role and elasticsearch_user module

### PR DESCRIPTION
**Pull request by tbauriedel**
First draft for a elasticsearch_role module.

With that module you can create, update and delete access roles inside elasticsearch.
Uses the elasticsearch API.

Example Usage:
```
    - name: Create role
      netways.elasticstack.elasticsearch_role:
        name: new-role
        cluster:
          - manage_own_api_key
          - delegate_pki
        indicies:
          - names:
              - foobar321
              - barfoo123
            privileges:
              - read
              - write
        state: present
        host: https://localhost:9200
        auth_user: elastic
        auth_pass: changeMe123!
        verify_certs: false
        ca_certs: /etc/elasticsearch/certs/http_ca.crt

```

### CODE REVIEW

**Comment by tbauriedel** [comment]
Example usage for the elasticsearch_user module:
```
- name: Create elasticsearch user 'new-user'
      netways.elasticstack.elasticsearch_user:
        name: new-user1
        fullname: New User
        password: changeMe123!
        email: new@user.de
        roles:
          - new-role1
          - logstash-writer
        enabled: true
        state: present
        host: https://localhost:9200
        auth_user: elastic
        auth_pass: changeMe321!
        verify_certs: false
```

**Comment by tbauriedel** [comment]
@widhalmt since that modules where requested from your side, could you please have a look at it? Maybe test them also.

**Comment by widhalmt** [comment]
I have local testing environments and I will put this change through a few tests. I'm the wrong person to review the code of the module (yet) but I can do integration tests and check everything around the module.

**Comment by lcndsmr** [comment]
Genuine question: why would we need a module which calls upon a role? I have never seen this before, is this commonly done?

**Comment by tbauriedel** [comment]
Inspiration for the two modules comes from [here](https://github.com/NETWAYS/ansible-collection-elasticstack/blob/main/roles/logstash/tasks/logstash-security.yml#L405-L472), for example.
Roles and users are repeatedly created in the collection via several tasks so that the needed component has a user in Elastic.
In order to be able to do this cleanly and not always require several tasks with requests, etc., these modules were requested.

With them you can easily create, update and delete them.

Sure, you could do this in your Kibana. But we have some known customers which do this all within the collection (or want to)

The whole thing has turned into a discussion of principles, as the idea of creating an internal Elastic user within the Logstash role is somewhat at odds with the idea of separating roles. However, this will be addressed later and assumed as a requirement.

**Review by lcndsmr — CHANGES_REQUESTED** [review]

File: plugins/module_utils/elasticsearch_role.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            pre_role = None
+
+        res = self.put()
+
+        if res.raw['role']['created'] == True:
+            self.result['changed'] = True
+            self.result['msg'] = self.role_name + " has been created"
+            return
+
+        if pre_role != None:
```
Comment (lcndsmr) [inline plugins/module_utils/elasticsearch_role.py:63]: You can un-nest here too, if you return when pro_role == None

The rest looks a lot better!

**Review by lcndsmr — CHANGES_REQUESTED** [review]
Code looks amazing, but it has to be testet now, and I don'Ät know how to withdraw approval :D

**Comment by tbauriedel** [comment]
I have already implemented a [workflow](https://github.com/NETWAYS/ansible-collection-elasticstack/actions/runs/7814309143/job/21315311813?pr=308) to test the modules.

@lcndsmr I guess you can remove the change request now. I have requested Thomas for review. As long as he dont approve it, it should be finde. Auto-merging is also not enabled for this PR.

@widhalmt would you please give us a last "okay it works"? :)

**Review by lcndsmr — APPROVED** [review]

File: docs/module-elasticsearch_role.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+* *name*: Name of your role (**Required**)
+* *cluster*: List of clusters
+* *indicies*: List of indicies
+  * *names*: List of names (**Required**)
+  * *privileges*: List of privileges (**Required**)
+* *state*: State of the role (Default: `present`)
+* *host*: API endpoint (**Required**)
+* *auth_user*: User to authenticate on the Elasticsearch API (**Required**)
+* *auth_pass*: Password for the given user (**Required**)
+* *verify_certs*: Verify certificates (Default: True)
```
Comment (widhalmt) [inline docs/module-elasticsearch_role.md:26]:
```suggestion
* *verify_certs*: Verify certificates (Default: `true`)
```
Just a style change.

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
 .cache
 *.swp
-__pycache__*
\ No newline at end of file
+__pycache__*
+.vscode
```
Comment (widhalmt) [inline .gitignore:4]:
```suggestion
.vscode
```
Missing `newline` at the end of the file

File: docs/module-elasticsearch_user.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+* *fullname*: Fullname of your user
+* *password*: Password for your user (**Required**)
+* *email*: Email for your user
+* *roles*: List of roles (**Required**)
+* *enabled*: Define wheter this user should be enabled (Default: `true`)
+* *state*: State of the role. `absent` to delete the user (Default: `present`)
+* *host*: API endpoint (**Required**)
+* *auth_user*: User to authenticate on the Elasticsearch API (**Required**)
+* *auth_pass*: Password for the given user (**Required**)
+* *verify_certs*: Verify certificates (Default: True)
```
Comment (widhalmt) [inline docs/module-elasticsearch_user.md:27]:
```suggestion
* *verify_certs*: Verify certificates (Default: `true`)
```
Style change

File: molecule/elasticsearch_test_modules/converge.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
+---
+# The workaround for arbitrarily named role directory is important because the git repo has one name and the role within it another
+# Found at: https://github.com/ansible-community/molecule/issues/1567#issuecomment-436876722
+- name: Converge
+  collections:
+    - netways.elasticstack
+  hosts: all
+  vars:
+    #elasticsearch_security: true # needed for tests of > 7 releases
```
Comment (widhalmt) [inline molecule/elasticsearch_test_modules/converge.yml:9]:
```suggestion
```
If we don't need it anymore, please remove it

File: molecule/elasticsearch_test_modules/prepare.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+          - openssl
+          - python3
+        update_cache: yes
+
+    - name: Install python module dependencies
+      ansible.builtin.pip:
+        name: "{{ item }}"
+      loop:
+        - elasticsearch
+        - certifi
```
Comment (widhalmt) [inline molecule/elasticsearch_test_modules/prepare.yml:23]: Do we need this module only for tests or also for production? Maybe it's missing from the docs?

File: plugins/module_utils/api.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+from elasticsearch import Elasticsearch
+import ssl
+
+class Api():
+    def new_client_basic_auth(host, auth_user, auth_pass, ca_certs, verify_certs) -> Elasticsearch:
+        ctx = ssl.create_default_context(cafile=ca_certs)
+        ctx.check_hostname = False
+        ctx.verify_mode = False
+        return Elasticsearch(hosts=[host], basic_auth=(auth_user, auth_pass), ssl_context=ctx, verify_certs=verify_certs)
```
Comment (widhalmt) [inline plugins/module_utils/api.py:15]:
```suggestion
        return Elasticsearch(hosts=[host], basic_auth=(auth_user, auth_pass), ssl_context=ctx, verify_certs=verify_certs)
```
Missing newline

**Review by widhalmt — CHANGES_REQUESTED** [review]
I found a few minor issues and had one extra question. Those aside: THANK YOU.

**Review by widhalmt — APPROVED** [review]
Thanks for the initial code. And the reaction to my picky requests. Approved.

**Pull request merged**